### PR TITLE
Implement #471

### LINKED ISSUE #471: Warn about overlapping match arm
Hello.
It would be nice to have a warning for the following code:

```rust
let x = 5;
match x {
    1 ... 10 => println!("1 ... 10"),
    5 ... 15 => println!("5 ... 15"),
    _ => (),
}
```

because the first two match arms are overlapping.

Thanks.

### PR DESCRIPTION
**Pull request by mcarton**
fixes #471

### CODE REVIEW

File: src/matches.rs [hunk 1]
```diff
@@ -71,101 +96,204 @@ impl LateLintPass for MatchPass {
...
-                    let template = match_template(cx, expr.span, source, "*", ex);
-                    span_lint(cx, MATCH_REF_PATS, expr.span, &format!(
-                        "instead of prefixing all patterns with `&`, you can dereference the \
-                         expression: `{}`", template));
-                }
+    }
+}
+
+fn check_overlapping_arms(cx: &LateContext, arms: &[Arm]) {
+    if arms.len() >= 2 {
```
Comment (Manishearth) [inline src/matches.rs]: We should only run this lint if the type of the match is an integral one. Type lookups are cheap, but collecting all the arms into a vector and all are not, so it's better to exit early and avoid all the churn if the match isn't of the correct type.

File: src/matches.rs [hunk 2]
```diff
@@ -209,3 +337,71 @@ fn match_template(cx: &LateContext,
         }
     }
 }
+
+fn overlaping<T>(ranges: &[SpannedRange<T>]) -> Option<(&SpannedRange<T>, &SpannedRange<T>)>
```
Comment (Manishearth) [inline src/matches.rs]: spelling

File: tests/compile-fail/matches.rs [hunk 3]
```diff
@@ -115,5 +115,34 @@ fn ref_pats() {
     }
 }
 
+fn overlapping() {
+    const FOO : u64 = 2;
+
+    match 42 {
+        0 ... 10 => println!("0 ... 10"), //~ERROR
```
Comment (Manishearth) [inline tests/compile-fail/matches.rs]: please keep at least the beginning of the error message after `ERROR`

**Comment by Manishearth** [comment]
Mostly looks good, but I'd like a second review from @llogiq too

File: src/matches.rs [hunk 4]
```diff
@@ -58,6 +64,25 @@ declare_lint!(pub MATCH_REF_PATS, Warn,
...
+///
+/// ```
+/// let x = 5;
+/// match x {
+///     1 ... 10 => println!("1 ... 10"),
+///     5 ... 15 => println!("5 ... 15"),
+///     _ => (),
+/// }
+/// ```
+declare_lint!(pub MATCH_OVERLAPPING_ARM, Warn,
```
Comment (llogiq) [inline src/matches.rs]: Very minor nit: I like curly braces better, because they don't require a semicolon.

File: src/matches.rs [hunk 5]
```diff
@@ -71,101 +96,205 @@ impl LateLintPass for MatchPass {
...
-            {
-                span_help_and_lint(cx, SINGLE_MATCH, expr.span,
-                                   "you seem to be trying to use match for destructuring a \
-                                    single pattern. Consider using `if let`",
-                                   &format!("try\nif let {} = {} {}",
-                                            snippet(cx, arms[0].pats[0].span, ".."),
-                                            snippet(cx, ex.span, ".."),
-                                            expr_block(cx, &arms[0].body, None, "..")));
-            }
+            check_single_match(cx, ex, arms, expr);
```
Comment (llogiq) [inline src/matches.rs]: Good idea to pull this into separate methods. Makes the code more readable.

File: src/matches.rs (hunk 4 above)
Comment (mcarton) [inline src/matches.rs]: Do you mean `declare_lint!{…}`? I did what the existing lints in that file did.
There seem to be no consistency in clippy (48 () vs. 39 {}), so it could be better to make another PR for that.

File: src/matches.rs [hunk 6]
```diff
@@ -209,3 +338,71 @@ fn match_template(cx: &LateContext,
...
+            (&Kind::Start(_, ra), &Kind::End(_, rb)) => if ra.node != rb.node { return Some((ra, rb)) },
+            (&Kind::End(a, _), &Kind::Start(b, _)) if a != b => (),
+            _ => return Some((&a.range(), &b.range())),
+        }
+    }
+
+    None
+}
+
+#[test]
```
Comment (llogiq) [inline src/matches.rs]: We have a `tests/` folder in our project. Tests should go there (and not all of them need to be compile tests).

**Comment by llogiq** [comment]
Only small nits to pick. I'm happy to merge once those are addressed.

**Comment by frewsxcv** [comment]
@mcarton FYI, there are merge conflicts

**Comment by mcarton** [comment]
@frewsxcv fixed, thanks.
@llogiq is there still any problem with this?

File: src/matches.rs (hunk 4 above)
Comment (llogiq) [inline src/matches.rs]: OK, I'll see if I find the time.

**Comment by llogiq** [comment]
No, I'm fine with merging. Good work here. :smile: Thank you.
